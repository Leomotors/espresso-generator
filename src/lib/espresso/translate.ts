function bitsToText(bits: string) {
  let text = "";

  for (let i = 0; i < bits.length; i++) {
    if (bits[i] === "1") {
      text += String.fromCharCode(65 + i);
    } else if (bits[i] === "0") {
      text += String.fromCharCode(65 + i) + "'";
    }
  }

  return text;
}

function countTerms(bits: string) {
  return bits.split("").filter((bit) => bit !== "-").length;
}

export function translate(text: string) {
  const lines = text.split("\n");

  let line = 5 - 1;

  const outputCircuit = {} as Record<number, string[]>;

  while (++line < lines.length) {
    if (lines[line].startsWith(".")) {
      break;
    }

    const [input, output] = lines[line].split(" ");

    for (let i = 0; i < output.length; i++) {
      if (output[i] === "1") {
        outputCircuit[i] ??= [];
        outputCircuit[i].push(input);
      }
    }
  }

  // Each output use 1 OR Gate if there is more than 1 terms
  const orGates = Object.values(outputCircuit).filter(
    (term) => term.length > 1,
  ).length;

  // Each term use AND Gate if input size more than 1
  const andGates = Object.values(outputCircuit)
    .map((terms) =>
      terms
        .map((t) => t.replaceAll("-", "").length)
        .reduce((prev, curr) => prev + (curr > 1 ? 1 : 0), 0),
    )
    .reduce((prev, curr) => prev + curr, 0);

  const wires =
    Object.values(outputCircuit)
      .map((terms) =>
        terms
          .map((term) => term.replaceAll("-", "").length)
          .reduce((prev, curr) => prev + curr, 0),
      )
      .reduce((prev, curr) => prev + curr, 0) +
    andGates +
    orGates;

  const outputText = Object.entries(outputCircuit)
    .map(
      ([key, value]) =>
        `${String.fromCharCode(97 + Number(key))} = ${value
          // This mutate outputCircuit
          .sort((a, b) => countTerms(a) - countTerms(b))
          .map(bitsToText)
          .join(" + ")}`,
    )
    .join("\n");

  return {
    raw: text,
    bits: outputCircuit,
    text: outputText,
    complexity: {
      orGates,
      andGates,
      wires,
    },
  };
}

export function translateMultiple(texts: string[]) {
  const translated = texts.map(translate);

  return {
    raw: translated
      .map((t) => t.raw)
      .map((r, index) =>
        r.replace(".ob a", ".ob " + String.fromCharCode(97 + index)),
      )
      .join("\n\n"),
    bits: translated
      .map((t) => Object.values(t.bits))
      .reduce((prev, curr, index) => ({ ...prev, [index]: curr }), {}),
    text: translated
      .map((t) => t.text)
      .map((t, index) => t.replace("a", String.fromCharCode(97 + index)))
      .join("\n"),
    complexity: translated
      .map((t) => t.complexity)
      .reduce(
        (prev, curr) => ({
          orGates: prev.orGates + curr.orGates,
          andGates: prev.andGates + curr.andGates,
          wires: prev.wires + curr.wires,
        }),
        { orGates: 0, andGates: 0, wires: 0 },
      ),
  };
}
