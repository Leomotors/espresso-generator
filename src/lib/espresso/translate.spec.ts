import { describe, it, expect } from "vitest";

import { translate } from "./translate";

describe("translate", () => {
  it("7 segment", () => {
    expect(
      translate(`.i 4
.o 7
.ilb A B C D
.ob a b c d e f g
.p 9
-0-0 1001100
-0-1 0110000
--10 1001100
-01- 0101001
-1-0 0010011
--11 1110000
--00 0110010
-101 1011011
1--- 1001011
.e`),
    ).toMatchSnapshot();
  });
});
