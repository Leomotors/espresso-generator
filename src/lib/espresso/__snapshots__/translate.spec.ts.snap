// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`translate > 7 segment 1`] = `
{
  "bits": {
    "0": [
      "1---",
      "-0-0",
      "--10",
      "--11",
      "-101",
    ],
    "1": [
      "-0-1",
      "-01-",
      "--11",
      "--00",
    ],
    "2": [
      "-0-1",
      "-1-0",
      "--11",
      "--00",
      "-101",
    ],
    "3": [
      "1---",
      "-0-0",
      "--10",
      "-01-",
      "-101",
    ],
    "4": [
      "-0-0",
      "--10",
    ],
    "5": [
      "1---",
      "-1-0",
      "--00",
      "-101",
    ],
    "6": [
      "1---",
      "-01-",
      "-1-0",
      "-101",
    ],
  },
  "complexity": {
    "andGates": 25,
    "orGates": 7,
    "wires": 91,
  },
  "raw": ".i 4
.o 7
.ilb A B C D
.ob a b c d e f g
.p 9
-0-0 1001100
-0-1 0110000
--10 1001100
-01- 0101001
-1-0 0010011
--11 1110000
--00 0110010
-101 1011011
1--- 1001011
.e",
  "text": "a = A + B'D' + CD' + CD + BC'D
b = B'D + B'C + CD + C'D'
c = B'D + BD' + CD + C'D' + BC'D
d = A + B'D' + CD' + B'C + BC'D
e = B'D' + CD'
f = A + BD' + C'D' + BC'D
g = A + B'C + BD' + BC'D",
}
`;
